File: backend/src/main/java/iuh/labbooking/service/groupjoinrequest/GroupJoinRequestService.java
package iuh.labbooking.service.groupjoinrequest;

import iuh.labbooking.dto.request.groupjoinrequest.UpdateJoinRequestStatusRequest;
import iuh.labbooking.dto.request.groupjoinrequest.BulkJoinRequestUpdate;
import iuh.labbooking.dto.response.base.PageResponse;
import iuh.labbooking.dto.response.groupjoinrequest.GroupJoinRequestDetailResponse;
import iuh.labbooking.dto.response.groupjoinrequest.GroupJoinRequestResponse;
import iuh.labbooking.dto.response.groupjoinrequest.SecureGroupJoinRequestResponse;
import iuh.labbooking.model.User;

import java.time.LocalDate;
import java.util.List;
import java.util.Set;

public interface GroupJoinRequestService {

        void createJoinRequests(User user, Set<Long> researchGroupIds, String joinMessage);

        void createJoinRequest(User user, Long researchGroupId, String joinMessage);

        GroupJoinRequestDetailResponse findRequestById(Long requestId);

        PageResponse<List<GroupJoinRequestResponse>> findMyJoinRequests(int page, int size, String status);

        PageResponse<List<SecureGroupJoinRequestResponse>> filterJoinRequestsForMyGroups(
                        int page, int size, String status, String keyword,
                        Long researchGroupId,
                        LocalDate fromDate,
                        LocalDate toDate);

        PageResponse<List<SecureGroupJoinRequestResponse>> filterJoinRequestsByAdmin(
                        int page, int size, String status, String keyword,
                        Long researchGroupId,
                        LocalDate fromDate,
                        LocalDate toDate);

        PageResponse<List<SecureGroupJoinRequestResponse>> findJoinRequestsForGroup(Long groupId, int page, int size,
                        String status, String keyword);

        SecureGroupJoinRequestResponse approveJoinRequest(Long requestId, UpdateJoinRequestStatusRequest request);

        SecureGroupJoinRequestResponse rejectJoinRequest(Long requestId, UpdateJoinRequestStatusRequest request);

        GroupJoinRequestResponse cancelJoinRequest(Long requestId);

        void bulkApproveJoinRequests(BulkJoinRequestUpdate request);

        void bulkRejectJoinRequests(BulkJoinRequestUpdate request);
}

File: frontend/src/services/group-join-request.service.ts
import http from '@/lib/http'
import type { ApiResponse, PageResponse } from '@/schemas/base.schema'
import type {
  CreateJoinRequestRequest,
  GroupJoinRequestResponse,
  SecureGroupJoinRequestResponse,
  UpdateJoinRequestStatusRequest,
  GroupJoinRequestDetailResponse
} from '@/schemas/group-join-request.schema'

export const groupJoinRequestService = {
  createJoinRequest: (data: CreateJoinRequestRequest) => http.post<ApiResponse<string>>('/group-join-requests', data),

  findMyJoinRequests: (params: { page: number; size: number; status?: string }) =>
    http.get<ApiResponse<PageResponse<GroupJoinRequestResponse[]>>>('/group-join-requests/my-requests', {
      params
    }),

  filterJoinRequestsForMyGroups: (params: {
    page: number
    size: number
    status?: string
    keyword?: string
    researchGroupId?: number
    fromDate?: string
    toDate?: string
  }) =>
    http.get<ApiResponse<PageResponse<SecureGroupJoinRequestResponse[]>>>('/group-join-requests/my-groups', {
      params
    }),

  filterAllJoinRequestsAdmin: (params: {
    page: number
    size: number
    status?: string
    keyword?: string
    researchGroupId?: number
    fromDate?: string
    toDate?: string
  }) =>
    http.get<ApiResponse<PageResponse<SecureGroupJoinRequestResponse[]>>>('/group-join-requests/admin', {
      params
    }),

  findJoinRequestsForGroup: (
    groupId: number,
    params: { page: number; size: number; status?: string; keyword?: string }
  ) =>
    http.get<ApiResponse<PageResponse<SecureGroupJoinRequestResponse[]>>>(`/group-join-requests/group/${groupId}`, {
      params
    }),

  findRequestById: (requestId: number) =>
    http.get<ApiResponse<GroupJoinRequestDetailResponse>>(`/group-join-requests/${requestId}`),

  approveJoinRequest: (requestId: number, data?: UpdateJoinRequestStatusRequest) =>
    http.patch<ApiResponse<SecureGroupJoinRequestResponse>>(`/group-join-requests/${requestId}/approve`, data),

  rejectJoinRequest: (requestId: number, data: UpdateJoinRequestStatusRequest) =>
    http.patch<ApiResponse<SecureGroupJoinRequestResponse>>(`/group-join-requests/${requestId}/reject`, data),

  cancelJoinRequest: (requestId: number) =>
    http.patch<ApiResponse<GroupJoinRequestResponse>>(`/group-join-requests/${requestId}/cancel`),

  bulkApproveJoinRequests: (requestIds: number[], responseNote?: string) =>
    http.patch<ApiResponse<void>>('/group-join-requests/actions/bulk-approve', {
      requestIds,
      responseNote
    }),

  bulkRejectJoinRequests: (requestIds: number[], responseNote?: string) =>
    http.patch<ApiResponse<void>>('/group-join-requests/actions/bulk-reject', {
      requestIds,
      responseNote
    })
}
